pykafka/producer.py

```diff
@@ -180,56 +331,128 @@ def _get_partition_msgs(partition_id, req):
             attempt += 1
             if attempt < self._max_retries:
                 time.sleep(self._retry_backoff_ms / 1000)
-                self._produce(to_retry, attempt)
+                # we have to call _produce here since the broker/partition
+                # target for each message may have changed
+                for _ in self._produce(to_retry, attempt):
+                    pass
             else:
                 raise ProduceFailureError('Unable to produce messages. See log for details.')
 
-    def _partition_messages(self, messages):
-        """Assign messages to partitions using the partitioner.
+    @raise_worker_exceptions
+    def _update_leaders(self):
+        """Ensure each message in each queue is in the queue owned by its
+            partition's leader
 
-        :param messages: Iterable of messages to publish.
-        :returns:        Generator of ((key, value), partition_id)
+        This function empties all broker queues, maps their messages to the
+        current leaders for their partitions, and enqueues the messages in
+        the appropriate queues.
         """
-        partitions = self._topic.partitions.values()
-        for message in messages:
-            if isinstance(message, basestring):
-                key = None
-                value = message
-            else:
-                key, value = message
-            value = str(value)
-            yield (key, value), self._partitioner(partitions, message).id
+        # empty queues and figure out updated partition leaders
+        new_queue_contents = defaultdict(list)
+        for owned_broker in self._owned_brokers.itervalues():
+            owned_broker.lock.acquire()
+            current_queue_contents = owned_broker.flush(0, acquire=False,
+                                                        release_inflight=True)
+            for kv, partition_id in current_queue_contents:
+                partition_leader = self._topic.partitions[partition_id].leader
+                new_queue_contents[partition_leader.id].append((kv, partition_id))
+        # retain locks for all brokers between these two steps
+        for owned_broker in self._owned_brokers.itervalues():
+            owned_broker.enqueue(new_queue_contents[owned_broker.broker.id],
+                                 self._block_on_queue_full,
+                                 acquire=False)
+            owned_broker.resolve_event_state()
+            owned_broker.lock.release()
 
-    def _produce(self, message_partition_tups, attempt):
-        """Publish a set of messages to relevant brokers.
+    @raise_worker_exceptions
+    def _wait_all(self):
+        """Block until all messages in flight are sent
```

Comment:
Just a thought, but "in flight" is an awkward term to use for these messages. When I see "in flight," I think "on the network on their way to the broker."  To me, this is more like "pending" messages since they're waiting to be sent. I can see how it cuts either way, so just a thought.